### PR TITLE
Modify HiDPI handling to correctly operate on Wayland

### LINKED ISSUE #3429: `scalefactor` does not scale mouse on MacOS
- [x] are you running newest version (version from docs) ?
- [x] can you reproduce the bug with a fresh environment ? (`]activate --temp; add Makie`)
```julia
(jl_WkIh9H) pkg> st
Status `/private/var/folders/n0/367fnn5x0rg8p83b918ltdph0000gn/T/jl_WkIh9H/Project.toml`
  [5789e2e9] FileIO v1.16.1
  [e9467ef8] GLMakie v0.9.1
  [ee78f7c6] Makie v0.20.1
```
- [x] What platform + GPU are you on?
```julia
In [9]: versioninfo()
Julia Version 1.9.4
Commit 8e5136fa297 (2023-11-14 08:46 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M2
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 4 virtual cores
Environment:
  JULIA_IMAGE_THREADS = 1
```

MWE:
```julia
using FileIO
using GLMakie
brain = load(assetpath("brain.stl"))

GLMakie.activate!(;scalefactor = 1) 
# GLMakie.activate!(;scalefactor = 2) #bug

fig = Figure()
ax1 = Axis3(fig[1, 1], aspect = :equal, title = "aspect = :equal")
ax2 = Axis3(fig[1, 2], aspect = :data, title = "aspect = :data")
for ax in [ax1, ax2]
    mesh!(ax, brain, color = :gray80)
end
DataInspector(fig)
fig
```
With `scalefactor=1`, the mouse can select the bounding box at the right place:
image

With `scalefactor=2`, the mouse needs to be in the **top left** quadrant to select the bounding box:
image

I can not reproduce it on Windows and have not tried it on Linux.

### PR DESCRIPTION
**Pull request by jmert**
Wayland's window and framebuffer sizing acts more like the OSX support that already exists, in that the window size is always specified in terms of a logical size rather than pixel size.

Rewrite some of the special-case behaviors to choose logic based on the GLFW platform constants rather than Julia's `Sys.is*` OS queries.

## Type of change

Delete options that do not apply:

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] Added an entry in CHANGELOG.md (for new features and breaking changes)

Fixes #3429

### CODE REVIEW

**Comment by jmert** [comment]
Note that to develop/test usage of Wayland, I also required locally-deployed versions of JuliaPackaging/Yggdrasil#9035, JuliaPackaging/Yggdrasil#9158, and JuliaPackaging/Yggdrasil#9160.

File: GLMakie/src/screen.jl [hunk 1]
```diff
@@ -346,7 +347,9 @@ function apply_config!(screen::Screen, config::ScreenConfig; start_renderloop::B
     ShaderAbstractions.switch_context!(glw)
     GLFW.SetWindowAttrib(glw, GLFW.FOCUS_ON_SHOW, config.focus_on_show)
     GLFW.SetWindowAttrib(glw, GLFW.DECORATED, config.decorated)
-    GLFW.SetWindowAttrib(glw, GLFW.FLOATING, config.float)
+    if GLFW.GetPlatform() != GLFW.PLATFORM_WAYLAND
+        GLFW.SetWindowAttrib(glw, GLFW.FLOATING, config.float)
```
Comment (SimonDanisch) [inline GLMakie/src/screen.jl:351]: Is floating not supported on Wayland?

File: GLMakie/src/events.jl [hunk 2]
```diff
@@ -170,11 +173,12 @@ end
...
     _, winh = window_size(nw)
-    @static if Sys.isapple()
+    if GLFW.GetPlatform() in (GLFW.PLATFORM_COCOA, GLFW.PLATFORM_WAYLAND)
+        sf = screen.scalefactor[] / scale_factor(nw)
         return w, (winh / sf) - h
     else
+        sf = screen.scalefactor[]
         return w / sf, (winh - h) / sf
     end
 end
```
Comment (ffreyer) [inline GLMakie/src/events.jl:184]: This isn't something that changed in this pr but the scaling with mac/wayland seems wrong to me. The GLFW docs state that GetCursorPos (`w, h`) and GetWindowSize (`window_size(nw)`) both return screen coordinates, so it should be `winh - h` not `winh / sf - h`. The scalefactor from GetContentScale (`scale_factor(nw)` should move that to pixel coordinates which I'm not sure we want here. Dropping the scaling of `w` here also seems odd.

Maybe you can check if
```julia
f, a, p = scatter(rand(10))
on(println, events(f).mouseposition)
display(f) # scalefactor = 2
```
gives you the same range of mouse coordinates as `f.scene.viewport[]` suggests within the window. For me on Windows this is always true, including with an OS level scale factor and a Makie level scale factor. If that's not the case you probably just want the else branch for Wayland here.

File: GLMakie/src/screen.jl (hunk 1 above)
Comment (jmert) [reply GLMakie/src/screen.jl:351]: Apparently not. If I leave that line in place, I get the following error.
```
julia> Figure()
Error showing value of type Figure:
ERROR: GLFWError (65548): Wayland: Platform does not support making a window floating
Stacktrace:
...
```
I haven't found any mention in GLFW's docs saying this isn't supported on Wayland, but given what I know of Wayland I'm also not surprised it's not a thing applications can do — the window manager instead has control of something like always-on-top behavior.

File: GLMakie/src/events.jl (hunk 2 above)
Comment (jmert) [reply GLMakie/src/events.jl:184]: I think the output indicates things are working as expected?
```

julia> display(f, scalefactor=2)
GLMakie.Screen(...)

julia> (0.9375, 0.3046875)
(0.9375, 0.84375)
(0.9375, 0.50390625)
(599.98828125, 449.3515625)
julia> f.scene.viewport[]
GeometryBasics.HyperRectangle{2, Int64}([0, 0], [600, 450])
```

The thing I tested when modifying these conditions was that click-and-dragging over a region to zoom in worked like expected, i.e. that I could select where I wanted to and the selection highlight followed my mouse.

But...I **do** have problems if I force the scalefactor to something other than the native 2 for my screen, so I think you're probably right that somethings not quite right. (I too was curious that `w` isn't scaled, but I was trying to limit my changes to just changing the platform-specific conditionals.)

File: GLMakie/src/events.jl [hunk 3]
```diff
@@ -170,11 +173,12 @@ end
...
-    sf = screen.scalefactor[] / (@static Sys.isapple() ? scale_factor(nw) : 1)
     _, winh = window_size(nw)
-    @static if Sys.isapple()
+    if GLFW.GetPlatform() in (GLFW.PLATFORM_COCOA, GLFW.PLATFORM_WAYLAND)
+        sf = screen.scalefactor[] / scale_factor(nw)
         return w, (winh / sf) - h
     else
+        sf = screen.scalefactor[]
         return w / sf, (winh - h) / sf
     end
```
Comment (jmert) [inline GLMakie/src/events.jl:183]:
```suggestion
    sf = screen.scalefactor[]
    if GLFW.GetPlatform() in (GLFW.PLATFORM_COCOA, GLFW.PLATFORM_WAYLAND)
        sf /= scale_factor(nw)
    end
    return w / sf, (winh - h) / sf
```
seems to work better for me when I change between scale factors 1, 2, and 3.

File: GLMakie/src/events.jl (hunk 2 above)
Comment (ffreyer) [reply GLMakie/src/events.jl]: Yea I think if the scalefactor matches the OS you just have a 1.0 multiplier here so the order doesn't matter.

I don't know whether this was correct on mac. I'd like to have that checked before merging this pr, but I don't have one to check on myself...

Comment (jmert) [reply GLMakie/src/events.jl]: A quick search of the issues for OSX mouse problems lead me to #3429, which reads to me like confirmation that the previous behavior wasn't working correctly on OSX. Maybe someone from that thread (@jacobleft or @pbouffard) can confirm that this branch either fixes the problem or at minimum doesn't make the situation even worse.

**Comment by SimonDanisch** [comment]
This PR is waiting for conformation whether #3429 is fixed in this PR on OSX:
```julia
using FileIO
using GLMakie
brain = load(assetpath("brain.stl"))

GLMakie.activate!(;scalefactor = 1) 
# GLMakie.activate!(;scalefactor = 2) # bug with mouseposition

fig = Figure()
ax1 = Axis3(fig[1, 1], aspect = :equal, title = "aspect = :equal")
ax2 = Axis3(fig[1, 2], aspect = :data, title = "aspect = :data")
for ax in [ax1, ax2]
    mesh!(ax, brain, color = :gray80)
end
DataInspector(fig)
fig
```

**Comment by jacobleft** [comment]
I get this error when trying to use this branch.
```julia
(test) pkg> precompile
Precompiling project...
  ✗ Makie
  ✗ GLMakie
  0 dependencies successfully precompiled in 33 seconds. 233 already precompiled.

ERROR: The following 2 direct dependencies failed to precompile:

GLMakie [e9467ef8-e4e7-5192-8a1a-b1aee30e663a]

Failed to precompile GLMakie [e9467ef8-e4e7-5192-8a1a-b1aee30e663a] to "/Users/jacob/.julia/compiled/v1.10/GLMakie/jl_fCFizV".
ERROR: LoadError: KeyError: key :stroke_depth_shift not found
Stacktrace:
  [1] getindex
    @ ./dict.jl:498 [inlined]
  [2] getindex
    @ ~/.julia/packages/MakieCore/f3yyf/src/attributes.jl:106 [inlined]
  [3] getindex(x::MakieCore.Poly{Tuple{GeometryBasics.HyperRectangle{2, Int64}}}, key::Symbol)
    @ MakieCore ~/.julia/packages/MakieCore/f3yyf/src/attributes.jl:192
  [4] getproperty
    @ ~/.julia/packages/MakieCore/f3yyf/src/attributes.jl:86 [inlined]
  [5] plot!(plot::MakieCore.Poly{Tuple{GeometryBasics.HyperRectangle{2, Int64}}})
    @ Makie ~/.julia/dev/Makie/src/basic_recipes/poly.jl:50
  [6] connect_plot!(parent::Makie.Scene, plot::MakieCore.Poly{Tuple{GeometryBasics.HyperRectangle{2, Int64}}})
    @ Makie ~/.julia/dev/Makie/src/interfaces.jl:380
  [7] plot!(scene::Makie.Scene, plot::MakieCore.Poly{Tuple{GeometryBasics.HyperRectangle{2, Int64}}})
    @ Makie ~/.julia/dev/Makie/src/interfaces.jl:389
  [8] _create_plot!(F::Function, attributes::Dict{Symbol, Any}, scene::Makie.Scene, args::Observables.Observable{GeometryBasics.HyperRectangle{2, Int64}})
    @ Makie ~/.julia/dev/Makie/src/figureplotting.jl:391
  [9] #poly!#91
    @ ~/.julia/packages/MakieCore/f3yyf/src/recipes.jl:440 [inlined]
 [10] poly!
    @ ~/.julia/packages/MakieCore/f3yyf/src/recipes.jl:438 [inlined]
 [11] initialize_block!(ax::Makie.Axis; palette::Nothing)
    @ Makie ~/.julia/dev/Makie/src/makielayout/blocks/axis.jl:191
 [12] initialize_block!(ax::Makie.Axis)
    @ Makie ~/.julia/dev/Makie/src/makielayout/blocks/axis.jl:150
 [13] _block(T::Type{Makie.Axis}, fig_or_scene::Makie.Figure, args::Vector{Any}, kwdict::Dict{Symbol, Any}, bbox::Nothing; kwdict_complete::Bool)
    @ Makie ~/.julia/dev/Makie/src/makielayout/blocks.jl:371
 [14] _block
    @ ~/.julia/dev/Makie/src/makielayout/blocks.jl:291 [inlined]
 [15] #_block#1420
    @ ~/.julia/dev/Makie/src/makielayout/blocks.jl:259 [inlined]
 [16] _block(::Type{Makie.Axis}, ::Makie.Figure)
    @ Makie ~/.julia/dev/Makie/src/makielayout/blocks.jl:258
 [17] _block(::Type{Makie.Axis}, ::GridLayoutBase.GridPosition; kwargs::@Kwargs{})
    @ Makie ~/.julia/dev/Makie/src/makielayout/blocks.jl:253
 [18] _block
    @ ~/.julia/dev/Makie/src/makielayout/blocks.jl:246 [inlined]
 [19] #_#1418
    @ ~/.julia/dev/Makie/src/makielayout/blocks.jl:237 [inlined]
 [20] Block
    @ ~/.julia/dev/Makie/src/makielayout/blocks.jl:236 [inlined]
 [21] macro expansion
    @ ~/.julia/dev/Makie/src/precompiles.jl:29 [inlined]
 [22] macro expansion
    @ ~/.julia/packages/PrecompileTools/L8A3n/src/workloads.jl:78 [inlined]
 [23] top-level scope
    @ ~/.julia/dev/Makie/src/precompiles.jl:27
 [24] include(mod::Module, _path::String)
    @ Base ./Base.jl:495
 [25] include(x::String)
    @ Makie ~/.julia/dev/Makie/src/Makie.jl:1
 [26] top-level scope
    @ ~/.julia/dev/Makie/src/Makie.jl:382
 [27] include
    @ ./Base.jl:495 [inlined]
 [28] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::String)
    @ Base ./loading.jl:2222
 [29] top-level scope
    @ stdin:3
in expression starting at /Users/jacob/.julia/dev/Makie/src/precompiles.jl:26
in expression starting at /Users/jacob/.julia/dev/Makie/src/Makie.jl:1
in expression starting at stdin:3
```

**Comment by ffreyer** [comment]
You usually need to dev MakieCore, Makie and GLMakie (or whichever backend you're using) to try prs. Otherwise you might end up desynched between the three.

**Comment by jacobleft** [comment]
@ffreyer Thanks.
Issue #3429 no longer reproduces. 
I can confirm it is fixed in this PR.

**Review by ffreyer — APPROVED** [review]
The test failures should be unrelated. I think the refimages are currently a bit mixed up

**Pull request merged**